### PR TITLE
Added core rl environment and implemented navigator with dqn and ddpg using Keras-RL

### PR DESCRIPTION
**Pull request by mhpanah**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (#913 #934 ) |
| Primary OS tested on | (Ubuntu 18.04) |
| Robotic platform tested on | (Gazebo + TB3) |

---

## Description of contribution in a few bullet points
- Modified TurtleBot3 RL Environment. Now there is:
  -  `Env` base class, which has all the needed functions for an RL environment based on gym
  - `GazeboInterface` class, which creates a node and has all the methods that are needed to communicate with gazebo and actuate the robot.
  - `Turtlebot3Environment` class, which has all the subs/pubs and basic methods that a turtlebot 3 needs for RL task.
  - `NavigationTaskEnv`, this class implements reward function and other functions that are needed for a navigation task.
- Added ddpg and dqn implementation for the navigator task which utilizes the keras-rl library.
- Modified the existing `dqn.py` to fix linter issues.  This file will soon be deleted as we are planning to use Kera-rl and Coach libraries. 

## Future work that may be required in bullet points
- Implement random_crawl similar to `navigator_dqn` and delete `dqn.py`, `parameters.py`, and `random_crawl_train.py`

### CODE REVIEW

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        while not self.get_entity_state.wait_for_service(timeout_sec=1.0):
+            print('get entity state service is not available...')
+        req = GetEntityState.Request()
+        req.name = 'turtlebot3_waffle'
+        future = self.get_entity_state.call_async(req)
+
+        while not future.done() and rclpy.ok():
+            sleep(0.01 / self.time_factor)
+
+        self.current_pose.position.x = future.result().state.pose.position.x
```
Comment (SteveMacenski) [inline nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]:
```
self.current_pose.position = future.result().state.pose.position
self.current_pose.orientation = future.result().state.pose.orientation 
```
is cleaner

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        if heading > pi:
+            heading -= 2 * pi
+
+        elif heading < -pi:
+            heading += 2 * pi
+
+        return heading
+
+    def get_yaw(self, q):
+        yaw = atan2(2.0 * (q.orientation.x * q.orientation.y + q.orientation.w * q.orientation.z),
```
Comment (SteveMacenski) [inline nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]: Is there really not a quaternion to yaw method in python to use?

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        self.current_pose = Pose()
+        self.goal_pose = Pose()
+
+        self.pub_cmd_vel = self.node_.create_publisher(Twist, 'cmd_vel', 1)
+        self.sub_scan = self.node_.create_subscription(LaserScan, 'scan', self.scan_callback,
+                                                       qos_profile_sensor_data)
+
+        self.reset_simulation = self.node_.create_client(Empty, 'reset_simulation')
+        self.reset_world = self.node_.create_client(Empty, 'reset_world')
+        self.unpause_proxy = self.node_.create_client(Empty, 'unpause_physics')
```
Comment (SteveMacenski) [inline nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]: I dont see you pausing the physics between runs. This is necessary to ensure the training is running at the correct rate with the correct propogation of physics between them.

Ei unpause -> act on environment -> let run for 1/frequency time -> pause -> return result. That way the simulation is stopping between updates and not continuing to creep.

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+import parameters
+
+import gym
+from keras.models import Sequential
+from keras.layers import Dense, Activation, Flatten
+from keras.optimizers import Adam
+from rl.agents.dqn import DQNAgent
+from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
+from rl.memory import SequentialMemory
```
Comment (SteveMacenski) [inline nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py]: I'm not familiar with this library, do you have a link?

Also are there rosdep keys for this? It should be added to the package.xml.

Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py]: Here is the link to the library: https://github.com/keras-rl/keras-rl.  You can use `pip install keras-rl` to instal. rosdep keys are there yet, I was actually planning to submit a PR to rosdistro to add it.

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py (hunk 2 above)
Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]: When I implemented this last month, the quaternion to yaw was not in python yet.  I need to check back to see if it is available now.

**Comment by mhpanah** [comment]
> Do you have a "convincing demo" of this work, working safely and reliably?
> 
> In quotes because I mean that in the loosest terms possible.

This is a work in progress.  I have a video demo with tb3 that can go from point A to B in an empty world.  However, the plan is to use the outcome of the global planner and just use this module for local planner.  This part is not implemented yet. Also, as the input, I'm planning to use local cosmap instead of raw laser data.

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py (hunk 4 above)
Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py]: Here is another library of interest that I'm planning to test. https://github.com/NervanaSystems/coach

**Comment by SteveMacenski** [comment]
Sounds similar to some work I've had going in the background. Maybe we should chat sometime on it. I've recently though started to shift away from DRL towards imitation leaning.

**Comment by mhpanah** [comment]
@SteveMacenski Yes, that sounds great. Let's chat sometime next week. Would Tuesday or Thursday work with your schedule?

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/gazebo_model/turtlebot3_waffle/model.sdf [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+          <noise>
+            <type>gaussian</type>
+            <mean>0.0</mean>
+            <stddev>0.01</stddev>
+          </noise>
+        </ray>
+        <plugin name="turtlebot3_laserscan" filename="libgazebo_ros_ray_sensor.so">
+          <ros>
+            <!-- <namespace>/tb3</namespace> -->
+            <argument>~/out:=scan</argument>
```
Comment (orduno) [inline nav2_experimental/nav2_rl/nav2_turtlebot3_rl/gazebo_model/turtlebot3_waffle/model.sdf]: Currently, you might need to add some prefixes for this to work, here's an [example](https://github.com/ros-planning/navigation2/pull/1118/files#diff-b0e13a3f492b823478a7c6e6eab8f52eR212)

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py (hunk 2 above)
Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]: Per this PR the tf2_ros is still not built for Python. https://github.com/ros2/geometry2/pull/99

Comment (SteveMacenski) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py]: Sounds good

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py (hunk 4 above)
Comment (SteveMacenski) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py]: I never found any implementations I liked so I ended up making my own, but I never ran across these two.

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/gazebo_model/turtlebot3_waffle/model.sdf (hunk 5 above)
Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/gazebo_model/turtlebot3_waffle/model.sdf]: @orduno I added the required `--ros-args` and `-r`. Thanks.

File: nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py (hunk 4 above)
Comment (mhpanah) [reply nav2_experimental/nav2_rl/nav2_turtlebot3_rl/navigator_dqn.py]: I have not verified these implementation, but since the second one is from Intel I'm planning to meet with them at some point and make it my default implementation.

**Comment by SteveMacenski** [comment]
Per our discussion, I think there should be an experimental branch as you suggested and that way you can roam free on this work and sync back up with us when its all done to add to the master branch. At that point we can split into several PRs (all the envs in 1, all the models in 1, etc), but I dont think at the stage this work is at it needs to be bogged down by reviews or architecture. Make something work and then we can worry about the rest
